File: services/healthreport/app/history.py
"""Trend series for the HTML report.

Reads the retained daily facts files so the email can show whether things are
getting better or worse, which a single day's counts cannot say.
"""

import json
import os
import re
from typing import Dict, List

FILENAME = re.compile(r"^facts-(\d{4})(\d{2})(\d{2})\.json$")


def trend(state_dir: str, days: int = 10) -> List[Dict]:
    """Return [{date, crit, warn}] oldest-first for the last `days` files."""
    history_dir = os.path.join(state_dir, "history")
    try:
        names = sorted(n for n in os.listdir(history_dir) if FILENAME.match(n))
    except OSError:
        return []

    out = []
    for name in names[-days:]:
        match = FILENAME.match(name)
        try:
            with open(os.path.join(history_dir, name)) as fh:
                facts = json.load(fh)
        except (OSError, ValueError):
            continue
        summary = facts.get("summary", {})
        out.append({
            "date": "%s-%s" % (match.group(2), match.group(3)),
            "crit": int(summary.get("crit", 0)),
            "warn": int(summary.get("warn", 0)),
        })
    return out


def by_host(facts: Dict) -> List[Dict]:
    """Actionable findings per subject, worst first.

    Subjects are hosts for most collectors but also repos and monitor names;
    that is the useful grouping - it answers "where is the trouble".
    """
    counts = {}
    for obs in facts.get("observations", []):
        severity = obs.get("severity")
        if severity not in ("crit", "warn"):
            continue
        entry = counts.setdefault(obs.get("subject") or "unknown", {"crit": 0, "warn": 0})
        entry[severity] += 1

    rows = [
        {"subject": subject, "crit": v["crit"], "warn": v["warn"], "total": v["crit"] + v["warn"]}
        for subject, v in counts.items()
    ]
    # Critical count dominates the ordering: three warnings are not worse than
    # one outage.
    rows.sort(key=lambda r: (-r["crit"], -r["total"], r["subject"]))
    return rows

File: tools/internal/healthreport/history.go
package healthreport

import (
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
)

var historyFilename = regexp.MustCompile(`^facts-(\d{4})(\d{2})(\d{2})\.json$`)

type TrendPoint struct {
	Date string `json:"date"`
	Crit int    `json:"crit"`
	Warn int    `json:"warn"`
}

// Trend returns [{date, crit, warn}] oldest-first for the last `days` files.
func Trend(stateDir string, days int) []TrendPoint {
	historyDir := filepath.Join(stateDir, "history")
	entries, err := os.ReadDir(historyDir)
	if err != nil {
		return nil
	}
	var names []string
	for _, e := range entries {
		if historyFilename.MatchString(e.Name()) {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)
	if len(names) > days {
		names = names[len(names)-days:]
	}

	var out []TrendPoint
	for _, name := range names {
		match := historyFilename.FindStringSubmatch(name)
		raw, err := os.ReadFile(filepath.Join(historyDir, name))
		if err != nil {
			continue
		}
		var facts struct {
			Summary struct {
				Crit int `json:"crit"`
				Warn int `json:"warn"`
			} `json:"summary"`
		}
		if err := json.Unmarshal(raw, &facts); err != nil {
			continue
		}
		out = append(out, TrendPoint{
			Date: match[2] + "-" + match[3],
			Crit: facts.Summary.Crit,
			Warn: facts.Summary.Warn,
		})
	}
	return out
}

type HostFindingCount struct {
	Subject string `json:"subject"`
	Crit    int    `json:"crit"`
	Warn    int    `json:"warn"`
	Total   int    `json:"total"`
}

// ByHost returns actionable findings per subject, worst first. Subjects are
// hosts for most collectors but also repos and monitor names; that is the
// useful grouping - it answers "where is the trouble".
func ByHost(facts *Facts) []HostFindingCount {
	type counts struct{ crit, warn int }
	byOrder := []string{}
	byCounts := map[string]*counts{}
	for _, obs := range facts.Observations {
		if obs.Severity != "crit" && obs.Severity != "warn" {
			continue
		}
		subject := obs.Subject
		if subject == "" {
			subject = "unknown"
		}
		c, ok := byCounts[subject]
		if !ok {
			c = &counts{}
			byCounts[subject] = c
			byOrder = append(byOrder, subject)
		}
		if obs.Severity == "crit" {
			c.crit++
		} else {
			c.warn++
		}
	}

	rows := make([]HostFindingCount, 0, len(byOrder))
	for _, subject := range byOrder {
		c := byCounts[subject]
		rows = append(rows, HostFindingCount{Subject: subject, Crit: c.crit, Warn: c.warn, Total: c.crit + c.warn})
	}
	// Critical count dominates the ordering: three warnings are not worse
	// than one outage.
	sort.SliceStable(rows, func(i, j int) bool {
		if rows[i].Crit != rows[j].Crit {
			return rows[i].Crit > rows[j].Crit
		}
		if rows[i].Total != rows[j].Total {
			return rows[i].Total > rows[j].Total
		}
		return rows[i].Subject < rows[j].Subject
	})
	return rows
}
